### PR TITLE
Force SILGen emission of on-demand synthesized accessors

### PR DESCRIPTION
**Pull request by slavapestov**
Fixes <rdar://problem/46503121>, <https://bugs.swift.org/browse/SR-9412>, and part of <rdar://problem/46186045>.

### CODE REVIEW

**Comment by slavapestov** [comment]
@swift-ci Please test

**Comment by slavapestov** [comment]
@swift-ci Please test source compatibility

**Comment by slavapestov** [comment]
@swift-ci Please smoke test

**Comment by slavapestov** [comment]
@swift-ci Please smoke test

**Comment by slavapestov** [comment]
@swift-ci Please test source compatibility

**Comment by slavapestov** [comment]
@swift-ci Please test source compatibility

**Review by rjmccall — COMMENTED** [review]
Seems reasonable, but one question.

File: lib/Sema/TypeChecker.cpp [hunk 1]
```diff
@@ -453,6 +453,17 @@ static void typeCheckFunctionsAndExternalDecls(SourceFile &SF, TypeChecker &TC)
...
+    // Synthesize any necessary function bodies.
+    // FIXME: If we're not planning to run SILGen, this is wasted effort.
+    while (!TC.FunctionsToSynthesize.empty()) {
+      auto function = TC.FunctionsToSynthesize.back().second;
+      TC.FunctionsToSynthesize.pop_back();
+      if (function.getDecl()->isInvalid() || TC.Context.hadError())
+        continue;
+
+      TC.synthesizeFunctionBody(function);
+    }
```
[the remark this answers is not in the document]
Comment (slavapestov) [reply lib/Sema/TypeChecker.cpp]: Imagine the accessor is on the external declarations list. `synthesizeFunctionBody()` doesn't type check the body, and `TC.typeCheckAbstractFunctionBody()` doesn't force the body to be synthesized.

However I have another PR that removes this altogether.

[the remark this answers is not in the document]
Comment (rjmccall) [reply lib/Sema/TypeChecker.cpp]: I don't understand.  This entire thing is inside an outer level of loop that should repeat it if any of the worklists are non-empty.

**Pull request merged**